Replace or fix `ord_subset` so `landmass` builds with `nightly`

TLDR; `ord_subset 3.1.`1 has a `?Sized` constraint in its tuple implementations that's incompatible with recent  `nightly` versions.

Long version: the crate does not compile with recent `nighly` any more and neither it nor its repo have been updated in 7+ years. There are alternatives like `ordered_float` or `decorum`, unless I miss sth. about it.

The alternative is to fix `ord_subset` and submit a PR upstream. But given the above maybe switching is the better option.

That said, the fix _is_ trivial:
* Remove the unnecessary `where last_type!($($T,)+): ?Sized` constraint.
* This constraint is redundant since tuple elements must be `Sized` anyway.

```
error[E0277]: the size for values of type `A` cannot be known at compilation time
   --> /home/ritz/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ord_subset-3.1.1/src/ord_subset_trait.rs:113:51
    |
113 |               impl<$($T:OrdSubset),+> OrdSubset for ($($T,)+) where last_type!($($T,)+): ?Sized {
    |                                                     ^^^^^^^^^ doesn't have a size known at compile-time
...
128 | / tuple_impls! {
129 | |     Tuple1 {
130 | |         (0) -> A
    | |                - this type parameter needs to be `Sized`
...   |
231 | | }
    | |_- in this macro invocation
    |
    = note: required for `(A,)` to implement `PartialEq`
note: required by a bound in `ord_subset_trait::OrdSubset`
   --> /home/ritz/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ord_subset-3.1.1/src/ord_subset_trait.rs:12:41
    |
12  | pub trait OrdSubset: PartialOrd<Self> + PartialEq<Self> {
    |                                         ^^^^^^^^^^^^^^^ required by this bound in `OrdSubset`
    = note: this error originates in the macro `tuple_impls` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider removing the `?Sized` bound to make the type parameter `Sized`
    |
113 -             impl<$($T:OrdSubset),+> OrdSubset for ($($T,)+) where last_type!($($T,)+): ?Sized {
113 +             impl<$($T:OrdSubset),+> OrdSubset for ($($T,)+)  {
    |
[snip ...]
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -17,7 +17,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        toolchain: [stable, beta]
+        toolchain: [stable, beta, nightly]
     steps:
       - uses: actions/checkout@v3
       - uses: dtolnay/rust-toolchain@stable

--- a/crates/landmass/Cargo.toml
+++ b/crates/landmass/Cargo.toml
@@ -17,7 +17,6 @@ keywords = ["navigation", "system", "pathfinding", "avoidance"]
 dodgy_2d = "0.5.4"
 glam = "0.29.1"
 kdtree = "0.7.0"
-ord_subset = "3.1.1"
 geo = "0.30.0"
 disjoint = "0.8.0"
 slotmap = "1.0.7"

--- a/crates/landmass/src/pathfinding.rs
+++ b/crates/landmass/src/pathfinding.rs
@@ -4,7 +4,6 @@ use std::{
 };
 
 use glam::Vec3;
-use ord_subset::OrdVar;
 
 use crate::{
   CoordinateSystem, Island, NavigationData, NodeType,
@@ -12,6 +11,7 @@ use crate::{
   nav_data::{BoundaryLinkId, NodeRef},
   nav_mesh::MeshEdgeRef,
   path::{BoundaryLinkSegment, IslandSegment, Path},
+  util::FloatOrd,
 };
 
 /// A concrete A* problem specifically for [`crate::Archipelago`]s.
@@ -291,8 +291,8 @@ pub(crate) fn find_path<CS: CoordinateSystem>(
         )
       })
       .filter(|pair| pair.1.is_finite())
-      .map(|pair| OrdVar::new_unchecked(pair.1))
-      .chain(std::iter::once(OrdVar::new_unchecked(1.0)))
+      .map(|pair| FloatOrd(pair.1))
+      .chain(std::iter::once(FloatOrd(1.0)))
       .min()
       .unwrap(),
     override_node_type_to_cost,

--- a/crates/landmass/src/util.rs
+++ b/crates/landmass/src/util.rs
@@ -1,9 +1,58 @@
-use std::mem::swap;
+use std::{cmp::Ordering, mem::swap, ops::Deref};
 
 use glam::{Quat, Vec3};
-use ord_subset::OrdVar;
 
 use crate::CoordinateSystem;
+
+/// An f32 that supports `Ord` operations.
+pub(crate) struct FloatOrd(pub(crate) f32);
+
+impl PartialEq for FloatOrd {
+  fn eq(&self, other: &Self) -> bool {
+    if self.0.is_nan() {
+      return other.0.is_nan();
+    }
+    self.0 == other.0
+  }
+}
+
+impl Eq for FloatOrd {}
+
+impl PartialOrd for FloatOrd {
+  #[inline]
+  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+    Some(self.cmp(other))
+  }
+}
+
+impl Ord for FloatOrd {
+  fn cmp(&self, other: &Self) -> Ordering {
+    // Pretend NaN is the biggest value.
+    if self.0.is_nan() {
+      return if other.0.is_nan() {
+        Ordering::Equal
+      } else {
+        Ordering::Greater
+      };
+    }
+    if self.0 > other.0 {
+      Ordering::Greater
+    } else if self.0 == other.0 {
+      Ordering::Equal
+    } else {
+      // This case also handles other=NaN since other is not greater and its not
+      // equal.
+      Ordering::Less
+    }
+  }
+}
+
+impl Deref for FloatOrd {
+  type Target = f32;
+  fn deref(&self) -> &Self::Target {
+    &self.0
+  }
+}
 
 /// A bounding box.
 #[derive(PartialEq, Clone, Copy, Debug)]
@@ -304,11 +353,11 @@ impl<ValueType> BoundingBoxHierarchy<ValueType> {
       .fold(BoundingBox::Empty, |acc, b| acc.expand_to_bounds(b));
     let bounds_size = bounding_box.size();
     if bounds_size.x > bounds_size.y && bounds_size.x > bounds_size.z {
-      values.sort_by_key(|v| OrdVar::new_unchecked(v.0.center().unwrap().x))
+      values.sort_by_key(|v| FloatOrd(v.0.center().unwrap().x))
     } else if bounds_size.y > bounds_size.z {
-      values.sort_by_key(|v| OrdVar::new_unchecked(v.0.center().unwrap().y))
+      values.sort_by_key(|v| FloatOrd(v.0.center().unwrap().y))
     } else {
-      values.sort_by_key(|v| OrdVar::new_unchecked(v.0.center().unwrap().z))
+      values.sort_by_key(|v| FloatOrd(v.0.center().unwrap().z))
     }
 
     let split_index = values.len() / 2;

